### PR TITLE
Optimize conda_initialize_content for fish-shell, improve the 400x speed

### PR DESCRIPTION
**Pull request by Mon-ius**
Hello conda maintainer team!

### Description

Recently, I have been switch from `zsh` to `fish`. And unsuprizedly, soon, I found out the performance issue during the `source config.fish`. According to my paste experience when uisng `bash` and `zsh`, I quickly target at the conda initilizaation process. But unfortunately, there is no much disscuss for fish yet. But luckily, I found the there existed a similar file located at `"%(conda_prefix)s/etc/fish/conf.d/conda.fish"`, which can just apply to fix the issues. So I researched a little about the conda  repo source code and put some little changes and hopefully, it can help more in the future!

### Checklist - did you ...

- [bug fixes / optimization] to func `_config_fish_content` under `conda/core/initialize.py`

### Performance Comparison

IMAGE 2023-06-20 14:55:13

### CODE REVIEW

**Comment by Mon-ius** [comment]
Hey,

I have been signed the `Conda Contributor License Agreement - Ver. 06.2021` (CLA).

Can we just keep moving on 🤗

**Comment by jezdez** [comment]
@Mon-ius I've left a comment about your CLA submission, https://github.com/conda/infrastructure/pull/784#issuecomment-1598523287

**Comment by jezdez** [comment]
@conda-bot check

**Comment by Mon-ius** [comment]
@jezdez Thanks for your information 🤗

**Comment by Mon-ius** [comment]
Seems like `tests/gateways/test_repodata_lock.py::test_lock_can_lock[False] - _queue.Empty` for `Tests / macos (conda-forge, 3.9, unit, 2, 2) (pull_request)` failed thus caused `Tests / Analyze results (pull_request)` failed.

I looked into the issued file `conda/tests/gateways/test_repodata_lock.py`, think may be the timeout issue for `test_lock_can_lock` func
```python
with env_vars(
        {"CONDA_PLATFORM": "osx-64", "CONDA_EXPERIMENTAL": "lock" if use_lock else ""},
        stack_callback=conda_tests_ctxt_mgmt_def_pol,
    ):
        cache = RepodataCache(tmp_path / "lockme", "repodata.json")

        qout = multiprocessing.Queue()  # put here, get in subprocess
        qin = multiprocessing.Queue()  # get here, put in subprocess

        p = multiprocessing.Process(target=locker, args=(cache, qin, qout))
        p.start()

        assert qin.get(timeout=6) == "ready"
        print("subprocess ready")

        with cache.cache_path_state.open("a+") as lock_file, lock(lock_file):
            print("lock acquired in parent process")
            qout.put("locked")
            if use_lock:
                assert isinstance(qin.get(timeout=13), OSError)
            else:
                assert qin.get(timeout=5) == "not locked"
            p.join(1)
            assert p.exitcode == 0
```

**Comment by Mon-ius** [comment]
@conda-bot check

**Comment by jezdez** [comment]
> @conda-bot check

No need to ask the conda-bot twice, it already checked if you've signed the CLA based on my review and assigned the `cla-signed` label

**Comment by Mon-ius** [comment]
@jezdez Thanks for explain for bot functionality,  now we have all test passed. 🤗

**Comment by Mon-ius** [comment]
shall we continue to merge the improvement according to the new analyze results 🤗

**Comment by beeankha** [comment]
Hi @Mon-ius ! Could you please rebase against the latest changes from `conda/main` and also add a [news file](https://github.com/conda/conda/tree/main/news) to this PR?

**Comment by Mon-ius** [comment]
Hi @beeankha

I have `rebase` the PR with the latest changes:

```bash
From github.com:Mon-ius/conda
   02f9878a5..d187c1968  main       -> origin/main
Updating 5a695c797..d187c1968
Fast-forward
 .github/workflows/stale.yml                               |    2 +-
 .gitignore                                                |    2 +-
 .pre-commit-config.yaml                                   |    4 +-
 .readthedocs.yml                                          |    2 +-
 conda/auxlib/packaging.py                                 |   82 +-
 conda/models/channel.py                                   |    5 +-
 conda/models/prefix_graph.py                              |    5 +-
 conda/models/records.py                                   |    5 +-
 conda/models/version.py                                   |    5 +-
 conda/plugins/__init__.py                                 |   25 +
 conda/plugins/hookspec.py                                 |   23 +-
 conda/plugins/manager.py                                  |   20 +-
 conda/plugins/types.py                                    |   38 +-
 conda/testing/integration.py                              |    2 +-
 docs/requirements.txt                                     |    4 +-
 docs/scrape_help.py                                       |  293 -------
 docs/source/{_ext => _extensions}/conda_umls.py           |    8 +-
 docs/source/_static/empty                                 |    0
 docs/source/_templates/autoapi/python/module.rst          |   95 +++
 docs/source/admin.rst                                     |    6 -
 docs/source/api/api.rst                                   |   20 -
 docs/source/api/index.rst                                 |   23 -
 docs/source/api/python_api.rst                            |   10 -
 docs/source/api/solver.rst                                |   11 -
 docs/source/changelog.rst                                 |    6 -
 docs/source/channels.rst                                  |    6 -
 docs/source/{commands.rst => commands/index.rst}          |    2 +-
 docs/source/conf.py                                       |  305 +++----
 docs/source/config.rst                                    |    6 -
 docs/source/custom-channels.rst                           |    6 -
 docs/source/dev-guide/api.rst                             |   15 +
 docs/source/dev-guide/index.rst                           |    2 +
 docs/source/dev-guide/plugins/index.rst                   |  140 ++--
 docs/source/dev-guide/plugins/manager.rst                 |   12 -
 docs/source/dev-guide/plugins/post_commands.rst           |    8 +-
 docs/source/dev-guide/plugins/pre_commands.rst            |    4 +-
 docs/source/dev-guide/plugins/solvers.rst                 |    4 +-
 docs/source/dev-guide/plugins/subcommands.rst             |    4 +-
 docs/source/dev-guide/plugins/virtual_packages.rst        |    4 +-
 docs/source/download.rst                                  |    6 -
 docs/source/env-commands.rst                              |    6 -
 docs/source/faq.rst                                       |    6 -
 docs/source/general-commands.rst                          |    6 -
 docs/source/get-started.rst                               |    6 -
 docs/source/help/conda-pip-virtualenv-translator.rst      |    6 -
 docs/source/help/silent.rst                               |    6 -
 docs/source/{ => img}/conda-logo.png                      |  Bin
 docs/source/index.rst                                     |    3 +-
 docs/source/install/central.rst                           |    6 -
 docs/source/install/full.rst                              |    6 -
 docs/source/install/quick.rst                             |    6 -
 docs/source/install/sample-condarc.rst                    |    6 -
 docs/source/install/tab-completion.rst                    |    6 -
 docs/source/installation.rst                              |    6 -
 docs/source/intro.rst                                     |    6 -
 docs/source/mro.rst                                       |    6 -
 docs/source/py2or3.rst                                    |    6 -
 docs/source/r-with-conda.rst                              |    6 -
 docs/source/redirects.rst                                 |   48 --
 docs/source/test-drive.rst                                |    6 -
 docs/source/travis.rst                                    |    6 -
 docs/source/troubleshooting.rst                           |    6 -
 docs/source/user-guide/tasks/use-conda-with-travis-ci.rst |    6 -
 docs/source/user-guide/tasks/use-mro-with-conda.rst       |    6 -
 docs/source/user-guide/tasks/use-r-with-conda.rst         |    6 -
 docs/source/using/cheatsheet.rst                          |    6 -
 docs/source/using/envs.rst                                |    6 -
 docs/source/using/index.rst                               |    6 -
 docs/source/using/pkgs.rst                                |    6 -
 docs/source/using/test-drive.rst                          |    6 -
 docs/source/using/using.rst                               |    6 -
 news/12798-add-sphinx-autoapi                             |   20 +
 news/12802-revise-plugins-index-page                      |   19 +
 news/12813-reenable-conda_env-tests                       |   19 +
 news/12819-convert-all-unittests-to-pytests               |   19 +
 recipe/meta.yaml                                          |    3 +-
 setup.cfg                                                 |   14 +-
 tests/base/test_context.py                                | 1050 ++++++++++++------------
 tests/common/test_configuration.py                        |  792 +++++++++---------
 tests/conda_env/installers/__init__.py                    |    2 -
 tests/conda_env/installers/test_pip.py                    |   55 --
 tests/conda_env/specs/test_base.py                        |   69 --
 tests/conda_env/specs/test_requirements.py                |   34 +-
 tests/conda_env/specs/test_yaml_file.py                   |   47 +-
 tests/conda_env/test_cli.py                               |  994 ++++++++++++-----------
 tests/conda_env/test_create.py                            |  449 +++++------
 tests/conda_env/test_env.py                               |  649 +++++++--------
 tests/core/test_envs_manager.py                           |  163 ++--
 tests/core/test_index.py                                  |   57 +-
 tests/core/test_path_actions.py                           |  881 ++++++++------------
 tests/core/test_portability.py                            |  229 +++---
 tests/core/test_prefix_data.py                            |  120 ++-
 tests/core/test_subdir_data.py                            |  221 +++--
 tests/gateways/disk/test_link.py                          |  132 ++-
 tests/gateways/test_connection.py                         |   82 +-
 tests/models/test_channel.py                              | 2176 +++++++++++++++++++++++++------------------------
 tests/models/test_dist.py                                 |    4 +-
 tests/models/test_index_record.py                         |  125 ++-
 tests/models/test_match_spec.py                           | 2212 +++++++++++++++++++++++++-------------------------
 tests/models/test_package_info.py                         |   87 +-
 tests/models/test_version.py                              |  826 +++++++++----------
 tests/test_activate.py                                    | 4543 ++++++++++++++++++++++++++++++++++++++++++++++++++++---------------------------------------------------
 tests/test_create.py                                      |   12 +-
 tests/test_exceptions.py                                  | 1093 +++++++++++++------------
 tests/test_export.py                                      |  154 ++--
 tests/test_fetch.py                                       |  141 ++--
 tests/test_import.py                                      |   57 --
 tests/test_install.py                                     |  640 ++++-----------
 tests/test_instructions.py                                |   89 +-
 tests/test_link_order.py                                  |   35 +-
 tests/test_plan.py                                        |  122 +--
 tests/test_priority.py                                    |  140 ++--
 tests/test_toposort.py                                    |  167 ++--
 tools/durations/Linux.json                                |  284 +++----
 tools/durations/Windows.json                              | 2469 +++++++++++++++++++++++++++++++-------------------------
 tools/durations/macOS.json                                | 2973 +++++++++++++++++++++++++++++++++++++------------------------------
 116 files changed, 12658 insertions(+), 13250 deletions(-)
```

Above indicates the details compared to my initial PR 

Waiting for GitHub workflow approved and all test passed 🤗

**Comment by Mon-ius** [comment]
Looked into the `macos (defaults, 3.9, integration, 1, 3) failed 7 hours ago in 37m 43s`.

It appears the network issues which caused the failure https://github.com/conda/conda/actions/runs/5426866990/jobs/9870770849?pr=12811#step:8:330:
```base
=================================== FAILURES ===================================
_________________ test_create_install_update_remove_smoketest __________________
Traceback (most recent call last):
  File "/Users/runner/work/conda/conda/tests/test_create.py", line 186, in test_create_install_update_remove_smoketest
    with make_temp_env("python=3.9") as prefix:
  File "/Users/runner/miniconda3/envs/conda-test-env/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/Users/runner/work/conda/conda/conda/testing/integration.py", line 322, in make_temp_env
    run_command(Commands.CREATE, prefix, *packages, **kwargs)
  File "/Users/runner/work/conda/conda/conda/deprecations.py", line 70, in inner
    return func(*args, **kwargs)
  File "/Users/runner/work/conda/conda/conda/testing/integration.py", line 288, in run_command
    result = do_call(args, p)
  File "/Users/runner/work/conda/conda/conda/cli/conda_argparse.py", line 134, in do_call
    result = getattr(module, func_name)(arguments, parser)
  File "/Users/runner/work/conda/conda/conda/notices/core.py", line 124, in wrapper
    return func(*args, **kwargs)
  File "/Users/runner/work/conda/conda/conda/cli/main_create.py", line 50, in execute
    install(args, parser, "create")
  File "/Users/runner/work/conda/conda/conda/cli/install.py", line 394, in install
    handle_txn(unlink_link_transaction, prefix, args, newenv)
  File "/Users/runner/work/conda/conda/conda/cli/install.py", line 421, in handle_txn
    unlink_link_transaction.download_and_extract()
  File "/Users/runner/work/conda/conda/conda/core/link.py", line 251, in download_and_extract
    self._pfe.execute()
  File "/Users/runner/work/conda/conda/conda/core/package_cache_data.py", line 855, in execute
    raise CondaMultiError(exceptions)
conda.CondaMultiError: ("Connection broken: ConnectionResetError(54, 'Connection reset by peer')", ConnectionResetError(54, 'Connection reset by peer'))
```

**Comment by Mon-ius** [comment]
I am aware of the robustness of test checks. 

What I modified is `fish` related part. However, another failed case [https://github.com/conda/conda/actions/runs/5426866990/jobs/9870770983?pr=12811] shows the `zsh` related failure. 

```bash
=========================== short test summary info ============================
FAILED tests/test_activate.py::test_zsh_basic_integration - pexpect.exceptions.TIMEOUT: <pexpect.popen_spawn.PopenSpawn object at 0x7feb6a70a0a0>
searcher: searcher_re:
    0: re.compile('Executing transaction: ...working... done.*\n')
<pexpect.popen_spawn.PopenSpawn object at 0x7feb6a70a0a0>
searcher: searcher_re:
    0: re.compile('Executing transaction: ...working... done.*\n')
= 1 failed, 63 passed, 1169 deselected, 1 xfailed, 1 xpassed, 129 warnings in 1799.26s (0:29:59) =
pytest-xprocess reminder::Be sure to terminate the started process by running 'pytest --xkill' if you have not explicitly done so in your fixture with 'xprocess.getinfo(<process_name>).terminate()'.
Error: Process completed with exit code 1.
```

**Comment by Mon-ius** [comment]
@beeankha looks like all checks passed, shall we move on now?

**Comment by travishathaway** [comment]
@Mon-ius,

We require that a new "news" file be created for each pull request. Can please copy the template located at `news/TEMPLATE` to a new file called `news/12811-optimize-fish-initialization` and then update the contents of that file with what you did? I would just add a single sentence underneath the "Enhancements" section.

Let us know if you have any questions.

**Comment by Mon-ius** [comment]
Hi @travishathaway, I just added that notes for it, thx for your kinda help! 🤗

File: news/12811-optimize-fish-initialization [hunk 1]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Optimize `_config_fish_content` function under `conda/core/initialize.py` for a 400x speed improvement.
```
Comment (beeankha) [inline news/12811-optimize-fish-initialization:3]:
```suggestion
* Optimize `_config_fish_content` function under `conda/core/initialize.py` for a 400x speed improvement. (#12811)
```

Comment (jezdez) [inline news/12811-optimize-fish-initialization]: I don't think end users are going to understand this, let's use this:

```suggestion
* Improve speed of fish shell initialization. (#12811)
```

**Review by jezdez — CHANGES_REQUESTED** [review]

File: news/12811-optimize-fish-initialization [hunk 2]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Optimize `_config_fish_content` function under `conda/core/initialize.py` for a 400x speed improvement. (#12811)
```
Comment (beeankha) [inline news/12811-optimize-fish-initialization:3]:
```suggestion
* Improve speed of fish shell initialization. (#12811)
```

**Review by beeankha — APPROVED** [review]

**Comment by beeankha** [comment]
Thank you so much for your contribution, @Mon-ius !

**Comment by Mon-ius** [comment]
Cheers 🍻  and thanks for all your time and help 🤗 @jezdez  @beeankha  @travishathaway

**Pull request merged**